prov/usnic/src/usdf_cq.c

```diff
@@ -119,7 +119,55 @@ usdf_cq_readerr(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
 
 	cq->cq_comp.uc_status = 0;
 
-	return sizeof(*entry);
+	return 1;
+}
+
+static ssize_t
+usdf_cq_readerr_soft(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
+		uint64_t flags)
+{
+	struct usdf_cq *cq;
+	size_t soft_cq_entry_len;
+	struct usdf_cq_soft_entry *soft_comp_entry;
+	void *tail;
+
+	USDF_TRACE_SYS(CQ, "\n");
+
+	cq = container_of(fcq, struct usdf_cq, cq_fid);
+
+	// The return values are analogous to sockets cq_readerr
+	tail = cq->c.soft.cq_tail;
+	soft_cq_entry_len = sizeof(struct usdf_cq_soft_entry);
+	soft_comp_entry = (struct usdf_cq_soft_entry *)tail;
+
+	entry->op_context = soft_comp_entry->op_context;
+	entry->flags = 0;
+	entry->err = FI_EIO;
+	switch (soft_comp_entry->status) {
+	case USD_COMPSTAT_SUCCESS:
+		entry->prov_errno = FI_SUCCESS;
+		break;
+	case USD_COMPSTAT_ERROR_CRC:
+		entry->prov_errno = FI_ECRC;
+		break;
+	case USD_COMPSTAT_ERROR_TRUNC:
+		entry->prov_errno = FI_ETRUNC;
+		break;
+	case USD_COMPSTAT_ERROR_TIMEOUT:
+		entry->prov_errno = FI_ETIMEDOUT;
+		break;
+	case USD_COMPSTAT_ERROR_INTERNAL:
+		entry->prov_errno = FI_EOTHER;
+		break;
+	}
+
+	tail = (uint8_t *)tail + soft_cq_entry_len;
```

Comment:
You can eliminate the `soft_cq_entry_len` variable and replace its only usage with `sizeof(struct usdf_cq_soft_entry)`.  Actually, better yet is to make both `tail` and `cq_tail` into a `struct usdf_cq_soft_entry *` instead of being a `void *`.  That way all the logic becomes type-safe and you can use simpler pointer arithmetic.